qa/L0_lifecycle/lifecycle_test.py

```diff
@@ -3326,6 +3326,27 @@ def test_model_config_overwite(self):
         updated_config = triton_client.get_model_config(model_name)
         self.assertEqual(original_config, updated_config)
 
+    def test_shutdown_while_background_unloading(self):
+        model_name = "identity_fp32"
+        triton_client = self._get_client()
+        self.assertTrue(triton_client.is_server_live())
+        self.assertTrue(triton_client.is_server_ready())
+        # Check the Python version of the model is loaded.
+        self.assertTrue(triton_client.is_model_ready(model_name, "1"))
+        python_model_config = triton_client.get_model_config(model_name)
+        self.assertEqual(python_model_config["backend"], "python")
+        # Load the Identity version, which will put the Python version into the
+        # background and unload it, the unload will take at least 10 seconds.
+        override_config = "{\n"
+        override_config += '"name": "identity_fp32",\n'
+        override_config += '"backend": "identity"\n'
+        override_config += "}"
+        triton_client.load_model(model_name, config=override_config)
+        identity_model_config = triton_client.get_model_config(model_name)
+        self.assertEqual(identity_model_config["backend"], "identity")
```

Comment:
Is doing all this any different than just calling `client.unload_model("identity_fp32")` ? Maybe I don't understand what the "background" models are.

Reply:
Good question. This is probably the most tricky behavior on the `model_lifecycle.cc`.

When a model is unloaded normally, it goes through the [`AsyncUnload()`](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L379) function, which keeps the model in the foreground. On the next `AsyncLoad()`, it will [reuse the object](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L433-L436). Therefore, the model never goes into the background if this route is taken.

When a model is reloaded (via `AsyncLoad()`), the [model is placed into the background](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L810-L816) if it cannot finish unloading by the time the reloaded model is ready (which should be the case for almost 100%).

Thus, the case is only reproducible when the model is reloaded and immediately followed by a server shutdown.

Reply:
A possible follow-up question: Why it is ok for the model to be unloading in the foreground while the server is shutting down?
When the model is unloading, the [`LiveModelStates()`](https://github.com/triton-inference-server/core/blob/main/src/model_repository_manager/model_lifecycle.cc#L188) will see the unloading model because its statis is neither unknown nor unavailable. Once it completes unloading, its status goes into unknown.

Reply:
Makes sense, thanks for explaining!

Just for my own curiosity - when the python model is put in background and identity model starts getting loaded -- does the python model immediately start unloading? Or only when the identity model has successfully finished loading, so we know that we don't need the background model for recovery anymore?

Reply:
Since we don't want any glitch during a reload, the Python model is left serving in the foreground while the identity model is loading in the background. Once the identity model finishes loading, the two models swap their places, so the Python model is in the background unloading and the identity model is in the foreground serving.